File: Median of the Subarrays.py
class Solution:
    def countSubarray(self, N, A, M): 
        def greaterMedian(m):
            tot = 0
            ans = 0
            pref = [0]*(2*N+1)
            cur = N
            pref[cur] += 1
            for i in A:
            
                x = -1
                if i >= m:
                    x = 1
                if x == 1:
                    tot += pref[cur]
                else:
                    tot -= pref[cur+x]
                ans += tot
                cur += x
                pref[cur] += 1
            return ans
        return greaterMedian(M) - greaterMedian(M+1)
                    

File: Median of the Subarrays.cpp
class Solution {
  public:
    long long solve(int &n,vector<int> &a, int m){
    int mp[2*n+1];
    memset(mp,0,sizeof(mp));
    long long cur=n,tot=0,ans=0;
    mp[cur]++;
    for(int i=0;i<n;i++){
      int x=-1;
      if(a[i]>=m)
        x=1;
      if(x==-1){
        tot-=mp[cur+x];
      }
      else{
        tot+=mp[cur];
      }
      cur+=x;
      ans+=tot;
      mp[cur]++;
    }
    return ans;
  }
 
    long long countSubarray(int N,vector<int> A,int M) {
        return solve(N,A,M)-solve(N,A,M+1);
    }
};
